squid-blockexplorer/src/types/events.ts

```diff
@@ -0,0 +1,297 @@
+    this._chain = ctx._chain
+    this.event = event
+  }
+
+  /**
+   * Some amount was removed from the account (e.g. for misbehavior).
+   */
+  get isV3(): boolean {
+    return this._chain.getEventHash('Balances.Slashed') === 'e84a34a6a3d577b31f16557bd304282f4fe4cbd7115377f4687635dc48e52ba5'
+  }
+
+  /**
+   * Some amount was removed from the account (e.g. for misbehavior).
+   */
+  get asV3(): {who: Uint8Array, amount: bigint} {
+    assert(this.isV3)
+    return this._chain.decodeEvent(this.event)
+  }
+}
+
+export class BalancesTransferEvent {
+  private readonly _chain: Chain
+  private readonly event: Event
+
+  constructor(ctx: EventContext)
+  constructor(ctx: ChainContext, event: Event)
+  constructor(ctx: EventContext, event?: Event) {
+    event = event || ctx.event
+    assert(event.name === 'Balances.Transfer')
+    this._chain = ctx._chain
+    this.event = event
+  }
+
+  /**
+   * Transfer succeeded.
+   */
+  get isV3(): boolean {
+    return this._chain.getEventHash('Balances.Transfer') === '0ffdf35c495114c2d42a8bf6c241483fd5334ca0198662e14480ad040f1e3a66'
+  }
+
+  /**
+   * Transfer succeeded.
+   */
+  get asV3(): {from: Uint8Array, to: Uint8Array, amount: bigint} {
+    assert(this.isV3)
+    return this._chain.decodeEvent(this.event)
+  }
+}
+
+export class BalancesUnreservedEvent {
+  private readonly _chain: Chain
+  private readonly event: Event
+
+  constructor(ctx: EventContext)
+  constructor(ctx: ChainContext, event: Event)
+  constructor(ctx: EventContext, event?: Event) {
+    event = event || ctx.event
+    assert(event.name === 'Balances.Unreserved')
+    this._chain = ctx._chain
+    this.event = event
+  }
+
+  /**
+   * Some balance was unreserved (moved from reserved to free).
+   */
+  get isV3(): boolean {
+    return this._chain.getEventHash('Balances.Unreserved') === 'e84a34a6a3d577b31f16557bd304282f4fe4cbd7115377f4687635dc48e52ba5'
+  }
+
+  /**
+   * Some balance was unreserved (moved from reserved to free).
+   */
+  get asV3(): {who: Uint8Array, amount: bigint} {
+    assert(this.isV3)
+    return this._chain.decodeEvent(this.event)
+  }
+}
+
+export class BalancesWithdrawEvent {
+  private readonly _chain: Chain
+  private readonly event: Event
+
+  constructor(ctx: EventContext)
+  constructor(ctx: ChainContext, event: Event)
+  constructor(ctx: EventContext, event?: Event) {
+    event = event || ctx.event
+    assert(event.name === 'Balances.Withdraw')
+    this._chain = ctx._chain
+    this.event = event
+  }
+
+  /**
+   * Some amount was withdrawn from the account (e.g. for transaction fees).
+   */
+  get isV3(): boolean {
+    return this._chain.getEventHash('Balances.Withdraw') === 'e84a34a6a3d577b31f16557bd304282f4fe4cbd7115377f4687635dc48e52ba5'
+  }
+
+  /**
+   * Some amount was withdrawn from the account (e.g. for transaction fees).
+   */
+  get asV3(): {who: Uint8Array, amount: bigint} {
+    assert(this.isV3)
+    return this._chain.decodeEvent(this.event)
+  }
+}
```

Comment:
should we do some class inheritance here? The distinct changes are the `event.name` and `EventHash`.

Reply:
this is autogenerated

Reply:
i do believe this is autogenerated with `typegen.json`.

Reply:
yes, I just tried and can confirm